### PR TITLE
Fix stateprep casting rules.

### PR DESCRIPTION
**Pull request by vincentmr**
### Before submitting

Please complete the following checklist when submitting a PR:

- [x] All new features must include a unit test.
      If you've fixed a bug or added code that should be tested, add a test to the
      [`tests`](../tests) directory!

- [x] All new functions and code must be clearly commented and documented.
      If you do make documentation changes, make sure that the docs build and
      render correctly by running `make docs`.

- [x] Ensure that the test suite passes, by running `make test`.

- [x] Add a new entry to the `.github/CHANGELOG.md` file, summarizing the
      change, and including a link back to the PR.

- [x] Ensure that code is properly formatted by running `make format`. 

When all the above are checked, delete everything above the dashed
line and fill in the pull request template.

------------------------------------------------------------------------------------------------------------

**Context:**
We currently have the following
```
@qml.qnode(qml.device("lightning.qubit", wires=2))
def circuit():
    qml.StatePrep(np.array([1, 0, 0, 0]), wires=[0, 1])
    return qml.state()

>>> [1 0 0 0]
```
why causes issues as the integer-valued state is not casted to complex.

**Description of the Change:**
Introduce ad hoc casting clause for integral types in Lightning.

**Benefits:**

**Possible Drawbacks:**

**Related GitHub Issues:**

### CODE REVIEW

**Review by AmintorDusko — APPROVED** [review]
Nice work! Thank you for that!

File: tests/test_comparison.py [hunk 1]
```diff
@@ -78,9 +78,7 @@ def circuit(measurement):
             basis state"""
             qml.BasisState(np.array(basis_state), wires=0)
             one_qubit_block(wires=0)
-            if callable(measurement):
-                return measurement()
-            return measurement
+            return measurement() if callable(measurement) else measurement
```
Comment (AmintorDusko) [inline tests/test_comparison.py:81]: Beautiful!

File: .github/CHANGELOG.md [hunk 2]
```diff
@@ -7,15 +7,18 @@
 
 ### Breaking changes
 
+* Cast integral-valued arrays to the device's complex type on entry in `_preprocess_state_vector` to make ensure the state is properly represented with floating-point numbers.  
+  [(#501)](https://github.com/PennyLaneAI/pennylane-lightning/pull/501)
```
Comment (AmintorDusko) [inline .github/CHANGELOG.md:11]:
```suggestion
* Cast integral-valued arrays to the device's complex type on entry in `_preprocess_state_vector` to ensure the state is correctly represented with floating-point numbers.  
  [(#501)](https://github.com/PennyLaneAI/pennylane-lightning/pull/501)
```

**Review by mlxd — APPROVED** [review]
LGTM -- thanks

**Pull request merged**